_posts/2020-03-16-dask-usage-at-blue-yonder.markdown

```diff
@@ -0,0 +1,161 @@
+uncompressed data in memory.
+
+### Downloading data from a database to a dataset
+
+One typical use case that we have is downloading large amounts of data from a database into a
+dataset. We use Dask/Distributed here as a means to parallelize the download and the graph is
+an extremely simple one: a lot of parallel nodes downloading pieces of data into a blob and one
+reduction node writing the dataset metadata (which is basically a dictionary what data can be found in which blob).
+
+### Doing computations on a dataset
+
+A lot of our algorithms for machine learning, forecasting, and optimization work on datasets.
+A source dataset is read, computations are performed on the data, and the result is written to
+a target dataset.
+In many cases, the layout of the source dataset (the partitioning, i.e., what data resides in which blob)
+is used for parallelization. This means the algorithms work independently on the individual
+blobs of the source dataset. Therefore, our respective Dask graphs again look very simple, with parallel nodes each performing
+the sequential operations of reading in the data from a source dataset blob, doing some computation on it,
+and writing it out to a target dataset blob. Again, there is a final reduction node writing the target
+dataset metadata. We typically use Dask's Delayed interface for these computations.
+
+![Simple Dask graph with parallel nodes and final reduction](/assets/images/2020-03-16-dask-usage-at-by-graph.png)
+
+### Dataset maintenance
+
+We use Dask/Distributed for performing maintenance operations like garbage collection on datasets.
+Garbage collection is necessary since we use a lazy approach when updating datasets and only delete or update
+references to blobs in the dataset metadata, deferring the deletion of no longer used blobs.
+In this case, Dask is used to parallelize the delete requests to the storage provider (Azure blob service).
+In a similar fashion, we use Dask to parallelize copy operations for backing up datasets.
+
+### Repartitioning a dataset
+
+Our algorithms rely on source datasets that are partitioned in a particular way. It is not always
+the case that we have the data available in a suitable partitioning, for instance, if the data results
+from a computation that used a different partitioning. In this case, we have to repartition the data.
+This works by reading the dataset as a [Dask dataframe](https://docs.dask.org/en/latest/dataframe.html) 
+repartitioning the dataframe using network shuffle, and writing it out again to a dataset.
+
+## Ad hoc data analysis
+
+Finally, our data scientists also use Dask for ad hoc analysis of data that is too large to
+fit into an individual compute node's memory.
+
+## Dask cluster setup at Blue Yonder
+
+We run a somewhat unique setup of Dask clusters that is partly driven by the specific requirements
+of our domain and partly historically grown. We are running a large number of individual clusters
```

Comment:
I would argue this is not historically grown but intentionally. Keyword: "environment isolation for SaaS applications" (And data isolation between customers) .This can be argued but this is how I motivated it when I was asked about it during the dask workshop